feat: degree is invariant under graph isomorphism

awueth opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks, and is labeled with `awaiting-review`.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

Mathlib has the definition `SimpleGraph.Iso.mapNeighborSet` which is an equivalence between neighbor sets induced by an isomorphism. Would it be beneficial to add the same equivalence for `neighborFinset`?

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

awueth committed:
feat: degree is invariant under graph isomorphism

semorrison commented:
If you look at the "PR summary" comment, and click on "Import changes for modified files", you'll see that this change significantly adds to the imports of this file. Can you refactor so this result goes in a different (possibly new?) file?

And yes, to the question in the PR description!

Thanks.

awueth commented:
I could move it to `Mathlib.Combinatorics.SimpleGraph.Finite` where `degree` lives but then I would need to import `Mathlib.Combinatorics.SimpleGraph.Maps` there. Is this better or should I create a new file?

semorrison commented:
My preference is a new file, e.g. `Map/Finite`, but I am more averse to imports than most. I'm happy to merge if you move it there, but also happy if another maintainer is happy to merge as is.

jcommelin reviewed (commented):
I suggest merging this, and we can split later, in an organized fashion.

awueth commented:
Apologies for not answering, I don't mind making a new file and adding the equivalent theorems for finite Neighbourhoods.

semorrison commented:
@awueth, up to you! Please add `awaiting-author` or `WIP` if you'd like to work on this more, and otherwise we'll keep reviewing as is.

urkud commented on Mathlib/Combinatorics/SimpleGraph/Maps.lean:
```diff
@@ -671,4 +672,17 @@ noncomputable def overFinIso (hc : Fintype.card V = n) : G ≃g G.overFin hc :=
+variable (v) [Fintype (G.neighborSet v)] (f : G ≃g G')
+
+instance : Fintype (G'.neighborSet (f v)) :=
+  Fintype.ofEquiv ↑(G.neighborSet v) (f.mapNeighborSet v)
```
This creates a diamond with the existing instance that gives `Fintype (G.neighborSet _)` from `Fintype G` and some `Decidable*`.

awueth replied:
What is a diamond and why is it problematic?

urkud commented on Mathlib/Combinatorics/SimpleGraph/Maps.lean:
```diff
@@ -671,4 +672,17 @@ noncomputable def overFinIso (hc : Fintype.card V = n) : G ≃g G.overFin hc :=
+instance : Fintype (G'.neighborSet (f v)) :=
+  Fintype.ofEquiv ↑(G.neighborSet v) (f.mapNeighborSet v)
+
+theorem degree_map_eq : G'.degree (f v) = G.degree v := by
```
IMHO, this should assume `Fintype` for both `neighborSet`s, so that it works for any instance.

awueth replied:
Should I remove the instance on line 679 then? Doesn't adding Fintype as an assumption for both `neighborSet`s make the theorem less general?

awueth committed:
move to new file

awueth commented:
I moved the contents of this pull request to a new file now and added `apply_mem_neighborFinset_iff` and `mapNeighborFinset`. I could create a `NeighborFinset` counterpart for any theorem in `SimpleGraph.Maps` about `NeighborSet`, but I'm not sure whether that is useful.
